blockchain/event_filter.go

```diff
@@ -222,3 +222,42 @@ func (e *EventFilter) appendBlockEvents(matchedEventsSofar []*FilteredEvent, hea
 	}
 	return matchedEventsSofar, processedEvents, nil
 }
+
+func (e *EventFilter) matchesEventKeys(eventKeys []*felt.Felt, keysMap []map[felt.Felt]bool) bool {
+	// short circuit if event doest have enough keys
+	for i := len(eventKeys); i < len(keysMap); i++ {
+		if len(keysMap[i]) > 0 {
+			return false
+		}
+	}
+
+	/// e.keys = [["V1", "V2"], [], ["V3"]] means:
+	/// ((event.Keys[0] == "V1" OR event.Keys[0] == "V2") AND (event.Keys[2] == "V3")).
+	//
+	// Essentially
+	// for each event.Keys[i], (len(e.keys[i]) == 0 OR event.Keys[i] is in e.keys[i]) should hold
+	for index, eventKey := range eventKeys {
+		// empty filter keys means match all
+		if index >= len(keysMap) || len(keysMap[index]) == 0 {
+			break
+		}
+		if found := keysMap[index][*eventKey]; !found {
+			return false
+		}
+	}
+
+	return true
+}
+
+func makeKeysMap(filterKeys [][]felt.Felt) []map[felt.Felt]bool {
+	filterKeysMaps := make([]map[felt.Felt]bool, len(filterKeys))
+	for index, keys := range filterKeys {
+		kMap := make(map[felt.Felt]bool, len(keys))
+		for _, key := range keys {
+			kMap[key] = true
+		}
+		filterKeysMaps[index] = kMap
+	}
+
+	return filterKeysMaps
+}
```

Comment:
If you are just interested in that something exists in a map it is better to use `struct{}` instead of `bool` (AFAIK).

```suggestion
func makeKeysMap(filterKeys [][]felt.Felt) []map[felt.Felt]struct{} {
	filterKeysMaps := make([]map[felt.Felt]struct{}, len(filterKeys))
	for index, keys := range filterKeys {
		kMap := make(map[felt.Felt]struct{}, len(keys))
		for _, key := range keys {
			kMap[key] = struct{}{}
		}
		filterKeysMaps[index] = kMap
	}

	return filterKeysMaps
}
```